### PR TITLE
feat: add binary_insertion_sort

### PR DESCRIPTION
**Pull request by Siddhartha-Padhy**
#### Description of Change
Sorting using Insertion sort algorithm with binary search.

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: Binary Insertion Sort

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1850"></a>

### CODE REVIEW

File: sorting/binary_insertion_sort.cpp [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+ * \f{bmatrix}{20 &30 &40 &50 &10\f}
+ * 4. We do not do anything with 50 and move on to the next iteration and
+ * select 10 which is misplaced and place it at correct position. Thus, we have
+ * \f{bmatrix}{10 &20 &30 &40 &50\f}
+ */
+
+#include <algorithm>
+#include <cassert>
+#include <iostream>
+#include <vector>
```
Comment (Panquesito7) [inline sorting/binary_insertion_sort.cpp:43]: Please add a one-line description of what the library/header is for (**see the example below**).

```c++
#include <iostream>    /// for IO operations
#include <cassert>    /// for assert
```

File: sorting/binary_insertion_sort.cpp [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+        arr[j + 1] = key;
+    }
+}
+} // namespace sorting
+
+/**
+ * @brief Self-test implementations
+ * @returns void
+ */
+void test() {
```
Comment (Panquesito7) [inline sorting/binary_insertion_sort.cpp:102]:
```suggestion
static void test() {
```

File: sorting/binary_insertion_sort.cpp [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+ * \brief Binary search function to find the most suitable pace for an element.
+ * \tparam T The generic data type.
+ * \param arr The actual vector in which we are searching a suitable place for the element.
+ * \param val The value for which suitable place is to be found.
+ * \param low The lower bound of the range we are searching in.
+ * \param high The upper bound of the range we are searching in.
+ * \returns the index of most suitable position of val.
+ */
+template<class T>
+int binary_search(std::vector<T> &arr,T val,int low,int high)
```
Comment (Panquesito7) [inline sorting/binary_insertion_sort.cpp:61]: Consider using `uint64_t` for non-negative values (or their appropriate size: `uint32_t`, `uint16_t`, `uint8_t`) or `int64_t` for negative values. Check other parts of the code ([reference](https://en.cppreference.com/w/cpp/types/integer)). Let us know if you need any help here or in our [**Discord server**](https://discord.gg/c7MnfGFGa6). 😃

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Great work! 😄👍
Please enable [GitHub Actions](https://docs.github.com/en/free-pro-team@latest/actions/learn-github-actions) in your repository of this fork in this link: https://github.com/Siddhartha-Padhy/C-Plus-Plus/actions

For any queries, comments, or questions, feel free to ask in our [**Discord server**](https://discord.gg/c7MnfGFGa6)! 🙂

**Comment by Siddhartha-Padhy** [comment]
All requested changes done!

File: sorting/binary_insertion_sort.cpp [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+ * \param val The value for which suitable place is to be found.
+ * \param low The lower bound of the range we are searching in.
+ * \param high The upper bound of the range we are searching in.
+ * \returns the index of most suitable position of val.
+ */
+template<class T>
+int64_t binary_search(std::vector<T> &arr,T val,int64_t low,int64_t high)
+{
+    if (high <= low)
+        return (val > arr[low]) ? (low + 1) : low;
```
Comment (Panquesito7) [inline sorting/binary_insertion_sort.cpp:64]:
```suggestion
    if (high <= low) {
        return (val > arr[low]) ? (low + 1) : low;
    }
```

File: sorting/binary_insertion_sort.cpp [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+    
+    int64_t mid = low + (high-low)/2;
+    if(arr[mid]>val)
+        return binary_search(arr,val,low,mid-1);
+    else if(arr[mid]<val)
+        return binary_search(arr,val,mid+1,high);
+    else
+        return mid+1;
```
Comment (Panquesito7) [inline sorting/binary_insertion_sort.cpp:72]:
```suggestion
    if(arr[mid]>val) {
        return binary_search(arr,val,low,mid-1);
    }
    else if(arr[mid]<val) {
        return binary_search(arr,val,mid+1,high);
    }
    else {
        return mid+1;
    }
```

File: sorting/binary_insertion_sort.cpp [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+/**
+ * @brief Self-test implementations
+ * @returns void
+ */
+static void test() {
+    /* descriptions of the following test */
+    /* 1st test: 
+       [5, -3, -1, -2, 7] returns [-3, -2, -1, 5, 7] */
+    std::vector<int64_t> arr1({5, -3, -1, -2, 7});
+    std::cout << "Test 1... ";
```
Comment (Panquesito7) [inline sorting/binary_insertion_sort.cpp:107]:
```suggestion
    std::cout << "1st test... ";
```

File: sorting/binary_insertion_sort.cpp [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+    sorting::insertionSort_binsrch(arr1);
+    assert(std::is_sorted(std::begin(arr1), std::end(arr1)));
+    std::cout << "passed" << std::endl;
+
+    /* 2nd test: 
+       [12, 26, 15, 91, 32, 54, 41] returns [12, 15, 26, 32, 41, 54, 91] */
+    std::vector<int64_t> arr2({12, 26, 15, 91, 32, 54, 41});
+    std::cout << "Test 2... ";
```
Comment (Panquesito7) [inline sorting/binary_insertion_sort.cpp:115]:
```suggestion
    std::cout << "2nd test... ";
```

File: sorting/binary_insertion_sort.cpp [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+    sorting::insertionSort_binsrch(arr2);
+    assert(std::is_sorted(std::begin(arr2), std::end(arr2)));
+    std::cout << "passed" << std::endl;
+
+    /* 3rd test: 
+       [7.1, -2.5, -4.0, -2.1, 5.7] returns [-4.0, -2.5, -2.1, 5.7, 7.1] */
+    std::vector<float> arr3({7.1, -2.5, -4.0, -2.1, 5.7});
+    std::cout << "Test 3... ";
```
Comment (Panquesito7) [inline sorting/binary_insertion_sort.cpp:123]:
```suggestion
    std::cout << "3rd test... ";
```

File: sorting/binary_insertion_sort.cpp [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+    sorting::insertionSort_binsrch(arr3);
+    assert(std::is_sorted(std::begin(arr3), std::end(arr3)));
+    std::cout << "passed" << std::endl;
+
+    /* 4th test: 
+       [12.8, -3.7, -20.7, -7.1, 2.2] returns [-20.7, -7.1, -3.7, 2.2, 12.8] */
+    std::vector<float> arr4({12.8, -3.7, -20.7, -7.1, 2.2});
+    std::cout << "Test 4... ";
```
Comment (Panquesito7) [inline sorting/binary_insertion_sort.cpp:131]:
```suggestion
    std::cout << "4th test... ";
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Almost there! 😄

**Comment by Siddhartha-Padhy** [comment]
Done!

**Review by Panquesito7 — APPROVED** [review]
LGTM 🚀 Thank you for your contribution! 😄👍

**Review by mishraabhinn — APPROVED** [review]

**Pull request merged**